lib/FusionInventory/Agent/Tools/Hardware.pm

```diff
@@ -540,18 +559,56 @@ sub _getMacAddress {
     # fallback on ports addresses (IF-MIB::ifPhysAddress) if unique
     my $addresses_oid = ".1.3.6.1.2.1.2.2.1.6";
     my $addresses = $snmp->walk($addresses_oid);
+
+    # filter keeping mac with a defined and associated ip otherwise takes all macs
+    my $ips = $snmp->walk('.1.3.6.1.2.1.4.20.1.2');
+    my @macs = ();
+    @macs = grep { defined } map { $addresses->{$_} } values %{$ips}
+        if (keys(%{$ips}));
+    @macs = grep { defined } values %{$addresses}
+        unless @macs;
+
     my @addresses =
         uniq
+        grep { /^$mac_address_pattern$/ }
         grep { $_ ne '00:00:00:00:00:00' }
         grep { $_ }
         map  { _getCanonicalMacAddress($_) }
-        values %{$addresses};
+        @macs;
```

Comment:
I know I'm the main culprit here, but the variable names are quite confusing, making reviewing the code quite painful:
- $addresses is a mapping of interface indexes => MAC addresses
- $ips is a mapping of IP addresses => interfaces indexes
- @macs is a list of mac addresses
Basically, they are all different sort of addresses, and the final conversion of @macs into @addresses is misleading.

I'd rather rename @macs as @all_mac_addresses, and @addresses as @valid_mac_addresses, for instances. Any other naming is fine, provided it show the final list is a subset of the same kind of object as the first one.